blockchain/store_test.go

```diff
@@ -0,0 +1,269 @@
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/tendermint/go-wire/data"
+	"github.com/tendermint/tendermint/blockchain"
+	"github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tmlibs/db"
+)
+
+func TestLoadBlockStoreStateJSON(t *testing.T) {
+	db := db.NewMemDB()
+
+	bsj := &blockchain.BlockStoreStateJSON{Height: 1000}
+	bsj.Save(db)
+
+	retrBSJ := blockchain.LoadBlockStoreStateJSON(db)
+
+	assert.Equal(t, *bsj, retrBSJ, "expected the retrieved DBs to match")
+}
+
+var blockStoreKey = []byte("blockStore")
+
+func TestNewBlockStore(t *testing.T) {
+	db := db.NewMemDB()
+	db.Set(blockStoreKey, []byte(`{"height": 10000}`))
+	bs := blockchain.NewBlockStore(db)
+	assert.Equal(t, bs.Height(), 10000, "failed to properly parse blockstore")
+
+	panicCausers := [][]byte{
+		[]byte("artful-doger"),
+		[]byte(""),
+		[]byte(" "),
+	}
+
+	for i, panicCauser := range panicCausers {
+		// Expecting a panic here on trying to parse an invalid blockStore
+		_, _, err := doFn(func() (interface{}, error) {
+			db.Set(blockStoreKey, panicCauser)
+			_ = blockchain.NewBlockStore(db)
+			return nil, nil
+		})
+		assert.NotNil(t, err, "#%d panicCauser: %q expected a panic", i, panicCauser)
+		// Note: We can't check for the exact substring because
+		// recovered stack traces as of Sept 2017 seem to
+		// mask the message. See https://play.golang.org/p/W9NTc4Fstr
+		// TODO: Perhaps file an error on the Go project.
+	}
+
+	db.Set(blockStoreKey, nil)
+	bs = blockchain.NewBlockStore(db)
+	assert.Equal(t, bs.Height(), 0, "expecting nil bytes to be unmarshaled alright")
+}
+
+func TestBlockStoreGetReader(t *testing.T) {
+	db := db.NewMemDB()
+	// Initial setup
+	db.Set([]byte("Foo"), []byte("Bar"))
+	db.Set([]byte("Foo1"), nil)
+
+	bs := blockchain.NewBlockStore(db)
+
+	tests := [...]struct {
+		key  []byte
+		want []byte
+	}{
+		0: {key: []byte("Foo"), want: []byte("Bar")},
+		1: {key: []byte("KnoxNonExistent"), want: nil},
+		2: {key: []byte("Foo1"), want: nil},
+	}
+
+	for i, tt := range tests {
+		r := bs.GetReader(tt.key)
+		if r == nil {
+			assert.Nil(t, tt.want, "#%d: expected a non-nil reader", i)
+			continue
+		}
+		slurp, err := ioutil.ReadAll(r)
+		if err != nil {
+			t.Errorf("#%d: unexpected Read err: %v", i, err)
+		} else {
+			assert.Equal(t, slurp, tt.want, "#%d: mismatch", i)
+		}
+	}
+}
+
+func freshBlockStore() (*blockchain.BlockStore, db.DB) {
+	db := db.NewMemDB()
+	return blockchain.NewBlockStore(db), db
+}
+
+func TestBlockStoreSaveLoadBlock(t *testing.T) {
+	bs, _ := freshBlockStore()
+	noBlockHeights := []int{0, -1, 100, 1000, 2}
+	for i, height := range noBlockHeights {
+		if g := bs.LoadBlock(height); g != nil {
+			t.Errorf("#%d: height(%d) got a block; want nil", i, height)
+		}
+	}
+
+	// Setup, test data
+	// The parts' data can be generated by running
+	// the code at https://gist.github.com/odeke-em/9ffac2b5df44595fad7084ece4c9bd98
+	part1 := &types.Part{Index: 0, Bytes: data.Bytes([]byte{
```

Comment:
i feel like we can simplify this test a bunch. do we really need these part sets? we can derrive part sets directly from blocks, and we have functions to make blocks so we shouldnt need much external madness. 

If we want incompelte/bad part sets, we can mutate them after derriving them from blocks.

Reply:
@ebuchman there are 3 different partsets that we are testing:
1) validAndCompletePartSet
2) incompletePartSet
3) UncontiguousPartSet

Mutating after deriving IMO makes it even more complex. Could you please help with an example of what you mean?

Reply:
For reference https://github.com/tendermint/tendermint/pull/677/files/23cb7d7c12ab48d30f23b2590c6bc3d3e72c9e29#diff-2dcc7092e28fd6d98ff9d9ef3b6bdcd1R118
<img width="622" alt="screen shot 2017-10-03 at 9 24 23 pm" src="https://user-images.githubusercontent.com/4898263/31158775-753d8f32-a881-11e7-9db5-60ff38e03448.png">

Reply:
cant we use block.MakePartSet ?